File: C++/17.迭代器模式/IteratorPatternDemo.cpp
/**
 * @file IteratorPatternDemo.cpp
 * @author {Layne} ({shu_huanglei@163.com})
 * @brief
 * @version 0.1
 * @date 2019-11-19
 *
 * @copyright Copyright (c) 2019
 *
 */

#include <iostream>

#include "NameRepository.h"

int main(int argc, char const *argv[]) {
    std::shared_ptr<NameRepository<std::string>> names_repository = std::make_shared<NameRepository<std::string>>();

    for (std::shared_ptr<Iterator<std::string>> iter = names_repository->GetIterator(); iter->HasNext();) {
        std::string name = iter->Next();
        std::cout << "Name: " << name << std::endl;
    }
    return 0;
}

File: Objective-C/17.IteratorPattern/IteratorPatternDemo.m
//
//  IteratorPatternDemo.m
//  17.IteratorPattern
//
//  Created by Layne on 2020/11/09.
//

#import "NameRepository.h"

int main(int argc, const char* argv[]) {
    @autoreleasepool {
        NSLog(@"Hello World!.");
        NameRepository* namesRepository = [[NameRepository alloc] init];

        for (id<IteratorProtocol> iter = [namesRepository getIterator];[iter hasNext];) {
            NSString* name = [iter next];
            NSLog(@"Name: %@", name);
        }
    }
    return 0;
}
